larpandoradlcontent/LArDeepLearning/DeepLearningTrackShowerIdAlgorithm.cc

```diff
@@ -0,0 +1,352 @@
+                if (inputEnergy < 0)
+                    continue;
+
+                const int pdg{std::abs(pMCParticle->GetParticleId())};
+                if (pdg == 11 || pdg == 22)
+                    tag = SHOWER;
+                else
+                    tag = TRACK;
+            }
+            catch (...)
+            {
+                continue;
+            }
+
+            featureVector.push_back(static_cast<double>(pCaloHit->GetPositionVector().GetX()));
+            featureVector.push_back(static_cast<double>(pCaloHit->GetPositionVector().GetZ()));
+            featureVector.push_back(static_cast<double>(tag));
+            featureVector.push_back(static_cast<double>(inputEnergy));
+        }
+        // Add number of hits to end of vector than rotate (more efficient than direct insert at front)
+        featureVector.push_back(static_cast<double>(featureVector.size() / 4));
+        std::rotate(featureVector.rbegin(), featureVector.rbegin() + 1, featureVector.rend());
+
+        LArMvaHelper::ProduceTrainingExample(trainingOutputFileName, true, featureVector);
+    }
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode DeepLearningTrackShowerIdAlgorithm::Infer()
+{
+    auto start = std::chrono::steady_clock::now();
+    LArDLHelper::TorchModel model;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, LArDLHelper::LoadModel(m_modelFileName, model));
+
+    if (m_visualize)
+        PANDORA_MONITORING_API(SetEveDisplayParameters(this->GetPandora(), true, DETECTOR_VIEW_XZ, -1.f, 1.f, 1.f));
+
+    for (const std::string listName : m_caloHitListNames)
+    {
+        const CaloHitList *pCaloHitList(nullptr);
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, listName, pCaloHitList));
+
+        const bool isU(pCaloHitList->front()->GetHitType() == TPC_VIEW_U ? true : false);
+        const bool isV(pCaloHitList->front()->GetHitType() == TPC_VIEW_V ? true : false);
+        const bool isW(pCaloHitList->front()->GetHitType() == TPC_VIEW_W ? true : false);
+
+        if (!isU && !isV && !isW)
+            return STATUS_CODE_NOT_ALLOWED;
+
+        const float tileSize = 128.f;
+        const int imageWidth = 256;
+        const int imageHeight = 256;
+        // Get bounds of hit region
+        float xMin{}; float xMax{}; float zMin{}; float zMax{};
+        this->GetHitRegion(*pCaloHitList, xMin, xMax, zMin, zMax);
+        const float xRange = xMax - xMin;
+        const float zRange = zMax - zMin;
+        int nTilesX = static_cast<int>(std::ceil(xRange / tileSize));
+        int nTilesZ = static_cast<int>(std::ceil(zRange / tileSize));
+        // Need to add 1 to number of tiles for ranges exactly matching tile size and for zero ranges
+        if (std::fmod(xRange, tileSize) == 0.f) ++nTilesX;
```

Comment:
Is this floating point comparison reliable?

Reply:
Yes, this is a very specific case in which we want this operation to be performed iff the floating point value returned is precisely 0.f.